File: python/numpy_version/collisions.py
import numpy as np
import constants as cs
from state import SimulationState
import math

def max_electron_coll_freq(sim: SimulationState) -> float:
    """
    Znajdowanie górnego limitu częstości kolizji dla elektronów (zvectoryzowane)
    """
    e = np.arange(cs.CS_RANGES, dtype=np.float64) * cs.DE_CS
    v = np.sqrt(2.0 * e * cs.EV_TO_J / cs.E_MASS)
    nu = v * sim.sigma_tot_e
    return float(np.max(nu))

def max_ion_coll_freq(sim: SimulationState) -> float:
    """
    Znajdowanie górnego limitu częstości kolizji dla jonów (zvectoryzowane)
    """
    e = np.arange(cs.CS_RANGES, dtype=np.float64) * cs.DE_CS
    g = np.sqrt(2.0 * e * cs.EV_TO_J / cs.MU_ARAR)
    nu = g * sim.sigma_tot_i
    return float(np.max(nu))

def collision_electron(sim: SimulationState, k: int, e_index: int):
    """
    Obsługa zderzenia elektron / Ar (przybliżenie zimnego gazu).
    Modyfikuje bezpośrednio stan symulacji (prędkości i liczbę cząstek).
    """
    # Pobranie aktualnych wartosci czastki
    xe: float = sim.x_e[k]
    vxe: float = sim.vx_e[k]
    vye: float = sim.vy_e[k]
    vze: float = sim.vz_e[k]

    # calculate relative velocity before collision & velocity of the centre of mass
    gx: float = vxe
    gy: float = vye
    gz: float = vze
    g: float = math.sqrt(gx * gx + gy * gy + gz * gz)
    wx: float = cs.F1 * vxe
    wy: float = cs.F1 * vye
    wz: float = cs.F1 * vze

    # find Euler angles
    if gx == 0.0:
        theta = 0.5 * cs.PI
    else:
        theta = math.atan2(math.sqrt(gy * gy + gz * gz), gx)

    if gy == 0.0:
        if gz > 0.0:
            phi = 0.5 * cs.PI
        else:
            phi = -0.5 * cs.PI
    else:
        phi = math.atan2(gz, gy)

    st = math.sin(theta)
    ct = math.cos(theta)
    sp = math.sin(phi)
    cp = math.cos(phi)

    # choose the type of collision based on the cross sections
    # take into account energy loss in inelastic collisions
    # generate scattering and azimuth angles
    # in case of ionization handle the 'new' electron

    t0 = sim.sigma[cs.E_ELA, e_index]
    t1 = t0 + sim.sigma[cs.E_EXC, e_index]
    t2 = t1 + sim.sigma[cs.E_ION, e_index]

    rnd = sim.rng.random()

    if rnd < (t0 / t2):
        # --- elastic scattering ---
        chi = math.acos(1.0 - 2.0 * sim.rng.random())   # isotropic scattering
        eta = cs.TWO_PI * sim.rng.random()  # azimuthal angle
    elif rnd < (t1 / t2):
        # --- excitation ---
        energy = 0.5 * cs.E_MASS * g * g                # electron energy
        energy = abs(energy - cs.E_EXC_TH * cs.EV_TO_J) # subtract energy loss
        g = math.sqrt(2.0 * energy / cs.E_MASS)         # relative velocity after energy loss
        chi = math.acos(1.0 - 2.0 * sim.rng.random())   # isotropic scattering
        eta = cs.TWO_PI * sim.rng.random()              # azimuthal angle
    else:
        # --- ionization ---
        energy = 0.5 * cs.E_MASS * g * g                # electron energy
        energy = abs(energy - cs.E_ION_TH * cs.EV_TO_J) # subtract energy loss
        
        # energy of the ejected electron
        e_ej = 10.0 * math.tan(sim.rng.random() * math.atan(energy / cs.EV_TO_J / 20.0)) * cs.EV_TO_J
        e_sc = abs(energy - e_ej)                              # energy of scattered electron
        
        g = math.sqrt(2.0 * e_sc / cs.E_MASS)           # relative velocity of scattered electron
        g2 = math.sqrt(2.0 * e_ej / cs.E_MASS)          # relative velocity of ejected electron
        
        chi = math.acos(math.sqrt(e_sc / energy))              # scattering angle for scattered electron
        chi2 = math.acos(math.sqrt(e_ej / energy))             # scattering angle for ejected electrons
        
        eta = cs.TWO_PI * sim.rng.random()              # azimuthal angle for scattered electron
        eta2 = eta + cs.PI                              # azimuthal angle for ejected electron
        
        sc = math.sin(chi2)
        cc = math.cos(chi2)
        se = math.sin(eta2)
        ce = math.cos(eta2)
        
        # Obliczenie prędkości dla wybitego elektronu 
        gx2 = g2 * (ct * cc - st * sc * ce)
        gy2 = g2 * (st * cp * cc + ct * cp * sc * ce - sp * sc * se)
        gz2 = g2 * (st * sp * cc + ct * sp * sc * ce + cp * sc * se)
        
        # Dodanie nowego elektronu na koniec listy
        sim.x_e[sim.N_e] = xe
        sim.vx_e[sim.N_e] = wx + cs.F2 * gx2
        sim.vy_e[sim.N_e] = wy + cs.F2 * gy2
        sim.vz_e[sim.N_e] = wz + cs.F2 * gz2
        sim.N_e += 1
        
        # Dodanie nowego jonu na koniec listy (prędkość z rozkładu termicznego tła)
        sim.x_i[sim.N_i] = xe
        sim.vx_i[sim.N_i] = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION)
        sim.vy_i[sim.N_i] = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION)
        sim.vz_i[sim.N_i] = sim.rng.normal(0.0, cs.NORMAL_DISTRIBUTION)
        sim.N_i += 1

    # scatter the primary electron
    sc = math.sin(chi)
    cc = math.cos(chi)
    se = math.sin(eta)
    ce = math.cos(eta)

    # compute new relative velocity
    gx_new = g * (ct * cc - st * sc * ce)
    gy_new = g * (st * cp * cc + ct * cp * sc * ce - sp * sc * se)
    gz_new = g * (st * sp * cc + ct * sp * sc * ce + cp * sc * se)
    
    # post-collision velocity of the colliding electron
    sim.vx_e[k] = wx + cs.F2 * gx_new
    sim.vy_e[k] = wy + cs.F2 * gy_new
    sim.vz_e[k] = wz + cs.F2 * gz_new


def collision_ion(sim: SimulationState, k: int, vx_a: float, vy_a: float, vz_a: float, e_index: int):
    """
    Obsługa zderzenia jon Ar+ / atom Ar.
    Modyfikuje bezpośrednio stan symulacji (prędkości wybranego jonu).
    """
    # pobranie aktualnych wlasciwosci jonu
    vx_1 = sim.vx_i[k]
    vy_1 = sim.vy_i[k]
    vz_1 = sim.vz_i[k]

    # calculate relative velocity before collision
    gx = vx_1 - vx_a
    gy = vy_1 - vy_a
    gz = vz_1 - vz_a
    g = math.sqrt(gx * gx + gy * gy + gz * gz)

    # velocity of the centre of mass
    wx = 0.5 * (vx_1 + vx_a)
    wy = 0.5 * (vy_1 + vy_a)
    wz = 0.5 * (vz_1 + vz_a)

    # find Euler angles
    if gx == 0.0:
        theta = 0.5 * cs.PI
    else:
        theta = math.atan2(math.sqrt(gy * gy + gz * gz), gx)
        
    if gy == 0.0:
        if gz > 0.0:
            phi = 0.5 * cs.PI
        else:
            phi = -0.5 * cs.PI
    else:
        phi = math.atan2(gz, gy)

    # determine the type of collision based on cross sections and generate scattering angle
    t1 = sim.sigma[cs.I_ISO, e_index]
    t2 = t1 + sim.sigma[cs.I_BACK, e_index]
    
    rnd = sim.rng.random()

    if rnd < (t1 / t2):
        # --- isotropic scattering ---
        chi = math.acos(1.0 - 2.0 * sim.rng.random())
    else:
        # --- backward scattering ---
        chi = cs.PI
        
    eta = cs.TWO_PI * sim.rng.random()
    
    sc = math.sin(chi)
    cc = math.cos(chi)
    se = math.sin(eta)
    ce = math.cos(eta)
    
    st = math.sin(theta)
    ct = math.cos(theta)
    sp = math.sin(phi)
    cp = math.cos(phi)

    # compute new relative velocity
    gx_new = g * (ct * cc - st * sc * ce)
    gy_new = g * (st * cp * cc + ct * cp * sc * ce - sp * sc * se)
    gz_new = g * (st * sp * cc + ct * sp * sc * ce + cp * sc * se)

    # post-collision velocity of the ion
    sim.vx_i[k] = wx + 0.5 * gx_new
    sim.vy_i[k] = wy + 0.5 * gy_new
    sim.vz_i[k] = wz + 0.5 * gz_new

File: Go/4.bce-loop-unrolling/collisions.go
package gopic

import (
	"math"
)

//----------------------------------------------------------------------//
// e / Ar collision  (cold gas approximation)                           //
//----------------------------------------------------------------------//

func (sim *SimulationState) CollisionElectron(xe float64, vxe, vye, vze *float64, eindex int) {
	// 1. Relative velocity before collision and center of mass velocity
	gx := *vxe
	gy := *vye
	gz := *vze
	g_perp_sq := gy*gy + gz*gz
	g_sq := gx*gx + g_perp_sq
	g := math.Sqrt(g_sq)
	g_perp := math.Sqrt(g_perp_sq)

	wx := F1 * (*vxe)
	wy := F1 * (*vye)
	wz := F1 * (*vze)

	// 2. Euler angles of initial velocity vector (vector algebra without Atan2/Sin/Cos)
	var ct, st, cp, sp float64
	if g > 0.0 {
		ct = gx / g
		st = g_perp / g
	} else {
		ct = 1.0
		st = 0.0
	}

	if g_perp > 0.0 {
		cp = gy / g_perp
		sp = gz / g_perp
	} else {
		cp = 1.0
		sp = 0.0
	}

	// 3. Select collision type (multiplicative selection without division)
	t0 := sim.Sigma[E_ELA][eindex]
	t1 := t0 + sim.Sigma[E_EXC][eindex]
	t2 := t1 + sim.Sigma[E_ION][eindex]
	rnd := sim.R01()
	r_t2 := rnd * t2

	eta := TWO_PI * sim.R01()
	se, ce := math.Sincos(eta)

	var sc, cc float64

	if r_t2 < t0 { // Elastic scattering (isotropic)
		cc = 1.0 - 2.0*sim.R01()
		sc = math.Sqrt(max(0.0, 1.0-cc*cc))
	} else if r_t2 < t1 { // Excitation (isotropic)
		energy := HALF_E_MASS * g_sq
		energy = math.Abs(energy - E_EXC_TH*EV_TO_J)
		g = math.Sqrt(energy * TWO_OVER_E_MASS)
		cc = 1.0 - 2.0*sim.R01()
		sc = math.Sqrt(max(0.0, 1.0-cc*cc))
	} else { // Ionization
		energy := HALF_E_MASS * g_sq
		energy = math.Abs(energy - E_ION_TH*EV_TO_J)

		// Energy of ejected electron (Opal's formula)
		e_ej := 10.0 * math.Tan(sim.R01()*math.Atan(energy*OPAL_FACTOR)) * EV_TO_J
		e_sc := math.Abs(energy - e_ej)

		g = math.Sqrt(e_sc * TWO_OVER_E_MASS)
		g2 := math.Sqrt(e_ej * TWO_OVER_E_MASS)

		cc = math.Sqrt(e_sc / energy)
		sc = math.Sqrt(max(0.0, 1.0-cc*cc))

		cc2 := math.Sqrt(e_ej / energy)
		sc2 := math.Sqrt(max(0.0, 1.0-cc2*cc2))

		// Azimuthal angle for ejected electron is eta + PI -> sin and cos negate
		se2 := -se
		ce2 := -ce

		gx2 := g2 * (ct*cc2 - st*sc2*ce2)
		gy2 := g2 * (st*cp*cc2 + ct*cp*sc2*ce2 - sp*sc2*se2)
		gz2 := g2 * (st*sp*cc2 + ct*sp*sc2*ce2 + cp*sc2*se2)

		// Add secondary electron
		sim.X_e[sim.N_e] = xe
		sim.Vx_e[sim.N_e] = wx + F2*gx2
		sim.Vy_e[sim.N_e] = wy + F2*gy2
		sim.Vz_e[sim.N_e] = wz + F2*gz2
		sim.N_e++

		// Add new ion
		sim.X_i[sim.N_i] = xe
		sim.Vx_i[sim.N_i] = sim.RMB()
		sim.Vy_i[sim.N_i] = sim.RMB()
		sim.Vz_i[sim.N_i] = sim.RMB()
		sim.N_i++
	}

	// 4. Scatter primary electron
	gx = g * (ct*cc - st*sc*ce)
	gy = g * (st*cp*cc + ct*cp*sc*ce - sp*sc*se)
	gz = g * (st*sp*cc + ct*sp*sc*ce + cp*sc*se)

	*vxe = wx + F2*gx
	*vye = wy + F2*gy
	*vze = wz + F2*gz
}

//----------------------------------------------------------------------//
// Ar+ / Ar collision                                                   //
//----------------------------------------------------------------------//

func (sim *SimulationState) CollisionIon(vx_1, vy_1, vz_1, vx_2, vy_2, vz_2 *float64, e_index int) {
	// 1. Process selection: isotropic vs backward scattering (charge exchange)
	t1 := sim.Sigma[I_ISO][e_index]
	t2 := t1 + sim.Sigma[I_BACK][e_index]
	rnd := sim.R01()

	if rnd*t2 >= t1 {
		// FAST-PATH: Backward scattering / Charge Exchange (I_BACK) - ~80% of ion collisions
		// The ion assumes target atom velocity directly without scattering geometry
		*vx_1 = *vx_2
		*vy_1 = *vy_2
		*vz_1 = *vz_2
		return
	}

	// 2. SLOW-PATH: Isotropic scattering (I_ISO) - ~20% of ion collisions
	gx := (*vx_1) - (*vx_2)
	gy := (*vy_1) - (*vy_2)
	gz := (*vz_1) - (*vz_2)
	g_perp_sq := gy*gy + gz*gz
	g_sq := gx*gx + g_perp_sq
	g := math.Sqrt(g_sq)
	g_perp := math.Sqrt(g_perp_sq)

	wx := 0.5 * ((*vx_1) + (*vx_2))
	wy := 0.5 * ((*vy_1) + (*vy_2))
	wz := 0.5 * ((*vz_1) + (*vz_2))

	var ct, st, cp, sp float64
	if g > 0.0 {
		ct = gx / g
		st = g_perp / g
	} else {
		ct = 1.0
		st = 0.0
	}

	if g_perp > 0.0 {
		cp = gy / g_perp
		sp = gz / g_perp
	} else {
		cp = 1.0
		sp = 0.0
	}

	cc := 1.0 - 2.0*sim.R01()
	sc := math.Sqrt(max(0.0, 1.0-cc*cc))

	eta := TWO_PI * sim.R01()
	se, ce := math.Sincos(eta)

	gx = g * (ct*cc - st*sc*ce)
	gy = g * (st*cp*cc + ct*cp*sc*ce - sp*sc*se)
	gz = g * (st*sp*cc + ct*sp*sc*ce + cp*sc*se)

	*vx_1 = wx + 0.5*gx
	*vy_1 = wy + 0.5*gy
	*vz_1 = wz + 0.5*gz
}
